src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs

```diff
@@ -1355,20 +1352,50 @@ public static PropertyFetch FetcherForProperty(Type? type, string propertyName)
                                 Log.Message($"Property {propertyName} is static.");
                                 return new PropertyFetch(type);
                             }
-                            Type typedPropertyFetcher = typeInfo.IsValueType ?
-                                typeof(ValueTypedFetchProperty<,>) : typeof(RefTypedFetchProperty<,>);
-                            Type instantiatedTypedPropertyFetcher = typedPropertyFetcher.GetTypeInfo().MakeGenericType(
-                                propertyInfo.DeclaringType!, propertyInfo.PropertyType);
-                            return (PropertyFetch)Activator.CreateInstance(instantiatedTypedPropertyFetcher, type, propertyInfo)!;
+
+                            return CreatePropertyFetch(typeInfo, propertyInfo);
                         }
                     }
 
+                    [UnconditionalSuppressMessage("AOT", "IL3050:RequiresDynamicCode",
+                        Justification = "MakeGenericType is only called when IsDynamicCodeSupported is true or only with ref types.")]
+                    private static PropertyFetch CreateEnumeratePropertyFetch(Type type, Type enumerableOfTType)
+                    {
+                        Type elemType = enumerableOfTType.GetGenericArguments()[0];
+#if NETCOREAPP
+                        if (!RuntimeFeature.IsDynamicCodeSupported && elemType.IsValueType)
```

Comment:
Would be better to have only `RuntimeFeature.IsDynamicCodeSupported` check and assert on `elemType.IsValueType` ?

Reply:
I don't believe there is a valid assert we could put on `elemType.IsValueType`. It is valid for it to be either `true` or `false`. (Note my test that I added is `IEnumerable<int>`.)

The reason I'm checking for `IsValueType` is so the behavior/implementation remains the same for `IEnumerable<RefType>` - since it is possible to still call `MakeGenericType` if the Type being passed is a ref type.